examples/read_pem_files/main.rs

```diff
@@ -0,0 +1,63 @@
+use config::{Config, Format, Map, Value, ValueKind, FileStoredFormat};
+
+/// Global settings for the exposing all preconfigured variables
+#[derive(serde::Deserialize, Clone, Debug)]
+pub struct Settings {
+    pub private_key: Option<String>,
+    pub public_key: Option<String>,
+}
+
+fn get_settings() -> Result<Settings, config::ConfigError> {
+    let settings = Config::builder()
+        // read public key
+        .add_source(config::File::new("examples/read_pem_files/cert/public.pem", PemFile).required(false))
+        // read private key
+        .add_source(config::File::new("examples/read_pem_files/cert/private.pem", PemFile).required(false))
+        .build()
+        .unwrap();
+
+    settings
+        .try_deserialize::<Settings>()
+}
+
+#[derive(Debug, Clone)]
+pub struct PemFile;
+
+impl Format for PemFile {
+    fn parse(
+        &self,
+        uri: Option<&String>,
+        text: &str,
+    ) -> Result<Map<String, config::Value>, Box<dyn std::error::Error + Send + Sync>> {
+
+        let mut result = Map::new();
+
+        let key = match text.contains("PRIVATE") {
+            true => String::from("private_key"),
+            false => String::from("public_key")
+        };
+
+        result.insert(
+            key,
+            Value::new(uri, ValueKind::String(text.into())),
+        );
```

Comment:
- Bit more verbosity in the example since documentation of implementing custom formats has been said to be lacking.
- I've improved your match statement, since the file contents could contain something other than a public key, it's better to match on the type found, otherwise throw an error (_also worth documenting an example of_).
- Instead of the `String::from()`, we can just handle this in the final map insert with `.to_owned()` which will convert `&str` to `String`.

```suggestion
        // Store any valid keys into this map, they'll be merged with other sources into the final config map:
        let mut result = Map::new();

        // Identify the PEM encoded data type by the first occurrence found:
        // NOTE: This example is kept simple, multiple or other encoded types are not handled.
        let key_type = ["PUBLIC", "PRIVATE"].into_iter().find(|s| text.contains(s));
        let key = match key_type {
            Some("PRIVATE") => "private_key",
            Some("PUBLIC") => "public_key",
            // Otherwise fail with an error message (the filename is implicitly appended):
            _ => return Err(Box::new(
                Error::new(ErrorKind::InvalidData, "PEM file did not contain a Private or Public key")
            )),
        };

        result.insert(
            key.to_owned(),
            Value::new(uri, ValueKind::String(text.into())),
        );
```

Reply:
I agree with you, this code works but it is not optimized for all types of .pem files and can result in unexpected behavior. My intention with this approach was just a simplification to have something quickly working.

Reply:
For some reason this line was give an error:
```let key_type = ["PUBLIC", "PRIVATE"].into_iter().find(|s| text.contains(s));```
I had to make this change in order to compile:
```let key_type = Vec::<&str>::from(["PUBLIC", "PRIVATE"]).into_iter().find(|s| text.contains(s));```